pyvisa/resources/async.py

```diff
@@ -0,0 +1,183 @@
+# -*- coding: utf-8 -*-
+"""
+    pyvisa.resources.async
+    ~~~~~~~~~~~~~~~~~~~~~~
+
+    Asynchronous communication for resources. This is kept as a separate module
+    because it uses the ``yield from`` syntax, which was only introduced in
+    Python 3.3. ``patch_class()`` is used to patch a given class with its async
+    methods.
+
+    This file is part of PyVISA.
+
+    :copyright: 2019 by PyVISA Authors, see AUTHORS for more details.
+    :license: MIT, see LICENSE for more details.
+"""
+import warnings
+from collections import defaultdict
+
+from .. import logger
+from .. import constants
+from .. import errors
+from ..compat import coroutine
+
+method_registry = defaultdict(list)
+
+
+def patch_class(cls):
+    """Patch a class with its async methods."""
+    for method in method_registry[cls.__name__]:
+        setattr(cls, method.__name__, method)
+
+
+def method_of(clsname):
+    def wrap(func):
+        method_registry[clsname].append(func)
+    return wrap
+
+
+@method_of('MessageBasedResource')
+def start_async_read(self, size=None):
+    """Start an asynchronous read operation and return immediately.
+
+    :param size: The chunk size to use when reading the data.
+
+    Only available for Python 3.3 and above.
+    """
+
+    if not self._io_events_enabled:
+        self.visalib.enable_event(self.session,
+                                  constants.VI_EVENT_IO_COMPLETION,
+                                  constants.VI_QUEUE)
+        self._io_events_enabled = True
+    size = self.chunk_size if size is None else size
+    buf, job_id, status = self.visalib.read_asynchronously(self.session, size)
+    self._jobs[job_id.value] = buf
+
+
+@method_of('MessageBasedResource')
+def check_async_read(self, timeout=0):
+    """Check if an async read is finished and return the message if so.
+
+    Returns None if the read operation has not yet finished. Note that this
+    method can raise a VisaIOError if the read operation itself times out. A
+    timeout of 0 can be used to poll the operation and return immediately.
+
+    :param timeout: How long (in milliseconds) to wait for the IO completion
+        event
+
+    Only available for Python 3.3 and above.
+    """
+    msg, _ = self._check_async_read(timeout)
+    return msg
+
+
+@method_of('MessageBasedResource')
+def _check_async_read(self, timeout=0):
+    try:
+        ev_type, ctx, status = self.visalib.wait_on_event(
+            self.session, constants.VI_EVENT_IO_COMPLETION, timeout)
+    except errors.VisaIOError as e:
+        if e.error_code == constants.VI_ERROR_TMO:
+            return None, None
+        raise
+
+    job_status, _ = self.visalib.get_attribute(ctx, constants.VI_ATTR_STATUS)
+    if job_status < 0:
+        raise errors.VisaIOError(job_status)
+
+    job_id, _ = self.visalib.get_attribute(ctx, constants.VI_ATTR_JOB_ID)
+    count, _ = self.visalib.get_attribute(ctx, constants.VI_ATTR_RET_COUNT)
+    self.visalib.close(ctx)
```

Comment:
I would prefer to see this in a try/finally block to be sure we always close the context.